flink-core/src/main/java/org/apache/flink/core/memory/DirectMemorySegment.java

```diff
@@ -0,0 +1,560 @@
+ * method implementation). Under this precondition, the JIT can perfectly inline methods.
+ * 
+ * This is harder to do and control with byte buffers, where different code paths use different versions of the class
+ * (heap, direct, mapped) and thus virtual method invocations are polymorphic and are not as easily inlined.
+ */
+public class DirectMemorySegment extends MemorySegment {
+	
+	/** The direct byte buffer that allocated the memory */
+	protected final ByteBuffer buffer;
+	
+	/** The address to the off-heap data */
+	private long address;
+	
+	/** The address one byte after the last addressable byte.
+	 *  This is address + size while the segment is not disposed */
+	private final long addressLimit;
+
+	/** The size in bytes of the memory segment */
+	private final int size;
+	
+	// -------------------------------------------------------------------------
+	//                             Constructors
+	// -------------------------------------------------------------------------
+
+	public DirectMemorySegment(int size) {
+		this(ByteBuffer.allocateDirect(size));
+	}
+
+	public DirectMemorySegment(ByteBuffer buffer) {
+		if (buffer == null || !buffer.isDirect()) {
+			throw new IllegalArgumentException();
+		}
+		
+		this.buffer = buffer;
+		this.size = buffer.capacity();
+		this.address = getAddress(buffer);
+		this.addressLimit = this.address + size;
+		
+		if (address >= Long.MAX_VALUE - Integer.MAX_VALUE) {
+			throw new RuntimeException("Segment initialized with too large address: " + address);
+		}
+	}
+
+	// -------------------------------------------------------------------------
+	//                        MemorySegment Accessors
+	// -------------------------------------------------------------------------
+	
+
+	@Override
+	public final boolean isFreed() {
+		return this.address > this.addressLimit;
+	}
+
+	public final void free() {
+		// this ensures we can place no more data and trigger
+		// the checks for the freed segment
+		this.address = this.addressLimit + 1;
+	}
+	
+	@Override
+	public final int size() {
+		return this.size;
+	}
+
+	@Override
+	public ByteBuffer wrap(int offset, int length) {
+		if (offset < 0 || offset > this.size || offset > this.size - length) {
+			throw new IndexOutOfBoundsException();
+		}
+		
+		this.buffer.limit(offset + length);
+		this.buffer.position(offset);
+		
+		return this.buffer;
+	}
+
+
+	// ------------------------------------------------------------------------
+	//                    Random Access get() and put() methods
+	// ------------------------------------------------------------------------
+
+	@Override
+	@SuppressWarnings("restriction")
+	public final byte get(int index) {
+		
+		final long pos = address + index;
+		if (index >= 0 && pos < addressLimit) {
+			return UNSAFE.getByte(pos);
+		}
+		else if (address > addressLimit) {
+			throw new IllegalStateException("disposed");
```

Comment:
I would actually vote to call the state `freed` and not disposed or rename the `free` method to `dispose`. This gives a better pointer to what it really means. ;)